src/pset/elip100.rs

```diff
@@ -0,0 +1,281 @@
+//! but excluding contract validation.
+//!
+//! ELIP0100 defines how to inlcude assets metadata, such as the contract defining the asset and
+//! the issuance prevout inside a PSET
+//!
+//! To use check [`PartiallySignedTransaction::add_asset_metadata`] and
+//! [`PartiallySignedTransaction::get_asset_metadata`]
+//!
+
+use std::io::Cursor;
+
+use super::{raw::ProprietaryKey, PartiallySignedTransaction};
+use crate::{
+    encode::{self, Decodable, Encodable},
+    AssetId, OutPoint,
+};
+
+/// keytype as defined in ELIP0100
+pub const PSBT_ELEMENTS_HWW_GLOBAL_ASSET_METADATA: u8 = 0x00u8;
+
+/// Prefix for PSET hardware wallet extension as defined in ELIP0100
+pub const PSET_HWW_PREFIX: &[u8] = b"pset_hww";
+
+/// Contains extension to add and retrieve from the PSET contract informations related to an asset
+impl PartiallySignedTransaction {
+    /// Add contract information to the PSET, returns None if it wasn't present or Some with the old
+    /// data if already in the PSET
+    pub fn add_asset_metadata(
+        &mut self,
+        asset_id: AssetId,
+        asset_meta: &AssetMetadata,
+    ) -> Option<Result<AssetMetadata, encode::Error>> {
+        let key = prop_key(&asset_id);
+        self.global
+            .proprietary
+            .insert(key, asset_meta.serialize())
+            .map(|old| AssetMetadata::deserialize(&old))
+    }
+
+    /// Get contract information from the PSET, returns None if there are no information regarding
+    /// the given `asset_id`` in the PSET
+    pub fn get_asset_metadata(
+        &self,
+        asset_id: AssetId,
+    ) -> Option<Result<AssetMetadata, encode::Error>> {
+        let key = prop_key(&asset_id);
+
+        self.global
+            .proprietary
+            .get(&key)
+            .map(|data| AssetMetadata::deserialize(data))
+    }
+}
+
+/// Asset metadata, the contract and the outpoint used to issue the asset
+#[derive(Debug, PartialEq, Eq)]
+pub struct AssetMetadata {
+    contract: String,
+    issuance_prevout: OutPoint,
+}
+
+fn prop_key(asset_id: &AssetId) -> ProprietaryKey {
+    let mut key = Vec::with_capacity(32);
+    asset_id
+        .consensus_encode(&mut key)
+        .expect("vec doesn't err"); // equivalent to asset_tag
+
+    ProprietaryKey {
+        prefix: PSET_HWW_PREFIX.to_vec(),
+        subtype: 0x00,
+        key,
+    }
+}
+
+impl AssetMetadata {
+    /// Returns the contract as string containing a json
+    pub fn contract(&self) -> &str {
+        &self.contract
+    }
+
+    /// Returns the issuance prevout where the asset has been issued
+    pub fn issuance_prevout(&self) -> OutPoint {
+        self.issuance_prevout
+    }
+
+    /// Serialize this metadata as defined by ELIP0100
+    ///
+    /// `<compact size uint contractLen><contract><32-byte prevoutTxid><32-bit little endian uint prevoutIndex>`
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut result = vec![];
+
+        self.contract
+            .as_bytes()
+            .to_vec()
```

Comment:
In 2ec896fefd8711f900c6222b48e71b34440b945a:

Tthis `to_vec` shouldn't be necessary -- I think we can directly consensus-encode byte slices.

Reply:
Yeah, I went lazy with vec for having the initial `VarInt`...
I see we have `consensus_encode_with_size` which is the right choice here, updated.